### PR TITLE
Migrate to Seccomp profile in security Context :warning:

### LINKED ISSUE #343: kubeaudit does not understand PodSecurityContext.seccompProfile
##### ISSUE TYPE
 - [ ] Bug Report
 
# BUG REPORT

##### SUMMARY

In k8s 1.20 was introduced GA syntax for setting seccompProfile:
```
    spec:
      securityContext:
        seccompProfile:
          type: RuntimeDefault
```

When running kubeaudit on pods with this setting, it complains about SeccompAnnotationMissing

##### ENVIRONMENT
* Kubeaudit version: v0.11
* Kubeaudit install method: docker image using "With RBAC" manifest

##### STEPS TO REPRODUCE

Deploy the kubernetes-dashboard according with https://github.com/kubernetes/dashboard/blob/master/aio/deploy/recommended.yaml, but the daemonset spec modified as follows:
```
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: kubernetes-dashboard
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        k8s-app: kubernetes-dashboard
    spec:
      containers:
      - args:
        - --auto-generate-certificates
        - --dashboard-endpoint=https://127.0.0.1:8443
        image: docker.io/kublr/k8s-dashboard-auth-proxy:v1.0
        imagePullPolicy: Always
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /
            port: 9443
            scheme: HTTPS
          initialDelaySeconds: 30
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 30
        name: kubernetes-dashboard-auth-proxy
        ports:
        - containerPort: 9443
          protocol: TCP
        resources:
          limits:
            cpu: 50m
            memory: 100Mi
          requests:
            cpu: 5m
            memory: 100Mi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          privileged: false
          readOnlyRootFilesystem: true
          runAsGroup: 2001
          runAsUser: 1001
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /certs
          name: kubernetes-dashboard-certs
        - mountPath: /tmp
          name: tmp-volume
      - args:
        - --auto-generate-certificates
        - --namespace=kubernetes-dashboard
        - --enable-skip-login=true
        image: docker.io/kubernetesui/dashboard:v2.0.4
        imagePullPolicy: Always
        name: kubernetes-dashboard
        ports:
        - containerPort: 8443
          protocol: TCP
        resources:
          limits:
            cpu: 100m
            memory: 512Mi
          requests:
            cpu: 10m
            memory: 50Mi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          privileged: false
          readOnlyRootFilesystem: true
          runAsGroup: 2001
          runAsUser: 1001
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /certs
          name: kubernetes-dashboard-certs
        - mountPath: /tmp
          name: tmp-volume
      dnsPolicy: ClusterFirst
      nodeSelector:
        kubernetes.io/os: linux
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        seccompProfile:
          type: RuntimeDefault
      serviceAccount: kubernetes-dashboard
      serviceAccountName: kubernetes-dashboard
      terminationGracePeriodSeconds: 30
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      volumes:
      - name: kubernetes-dashboard-certs
        secret:
          defaultMode: 420
          secretName: kubernetes-dashboard-certs
      - emptyDir: {}
        name: tmp-volume
```

Run kubeaudit by applying the following manifest:

```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: kubeaudit
  namespace: default

---

kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: kubeaudit
rules:
  - apiGroups: [""]
    resources:
      - pods
      - podtemplates
      - replicationcontrollers
      - namespaces
      - serviceaccounts
    verbs: ["list"]
  - apiGroups: ["apps"]
    resources:
      - daemonsets
      - statefulsets
      - deployments
    verbs: ["list"]
  - apiGroups: ["batch"]
    resources:
      - cronjobs
    verbs: ["list"]
  - apiGroups: ["networking.k8s.io"]
    resources:
      - networkpolicies
    verbs: ["list"]

---

kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: kubeaudit
subjects:
  - kind: ServiceAccount
    name: kubeaudit
    namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kubeaudit

---

apiVersion: batch/v1
kind: Job
metadata:
  name: kubeaudit
  namespace: default
spec:
  template:
    metadata:
      annotations:
        container.apparmor.security.beta.kubernetes.io/kubeaudit: runtime/default
        seccomp.security.alpha.kubernetes.io/pod: runtime/default
    spec:
      serviceAccountName: kubeaudit
      restartPolicy: OnFailure
      containers:
        - name: kubeaudit
          image: shopify/kubeaudit:v0.11
          args: ["all", "--exitcode", "0"]
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop: ["all"]
            privileged: false
            readOnlyRootFilesystem: true
            runAsNonRoot: true
```

##### EXPECTED RESULTS

Kubeaudit complains about AppArmorAnnotationMissing but not about SeccompAnnotationMissing.

##### ACTUAL RESULTS

```
---------------- Results for ---------------

  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: kubernetes-dashboard
    namespace: kubernetes-dashboard

--------------------------------------------

-- [error] AppArmorAnnotationMissing
   Message: AppArmor annotation missing. The annotation 'container.apparmor.security.beta.kubernetes.io/kubernetes-dashboard-auth-proxy' should be added.
   Metadata:
      Container: kubernetes-dashboard-auth-proxy
      MissingAnnotation: container.apparmor.security.beta.kubernetes.io/kubernetes-dashboard-auth-proxy

-- [error] AppArmorAnnotationMissing
   Message: AppArmor annotation missing. The annotation 'container.apparmor.security.beta.kubernetes.io/kubernetes-dashboard' should be added.
   Metadata:
      Container: kubernetes-dashboard
      MissingAnnotation: container.apparmor.security.beta.kubernetes.io/kubernetes-dashboard

-- [error] RunAsNonRootPSCNilCSCNil
   Message: runAsNonRoot is not set in container SecurityContext nor the PodSecurityContext. It should be set to 'true' in at least one of the two.
   Metadata:
      Container: kubernetes-dashboard-auth-proxy

-- [error] RunAsNonRootPSCNilCSCNil
   Message: runAsNonRoot is not set in container SecurityContext nor the PodSecurityContext. It should be set to 'true' in at least one of the two.
   Metadata:
      Container: kubernetes-dashboard

-- [error] SeccompAnnotationMissing
   Message: Seccomp annotation is missing. The annotation seccomp.security.alpha.kubernetes.io/pod: runtime/default should be added.
   Metadata:
      MissingAnnotation: seccomp.security.alpha.kubernetes.io/pod
```

### PR DESCRIPTION
**Pull request by Ser87ch**
##### Description

This PR changes `seccomp` auditor to scan `seccompProfile` instead of annotations as requested in https://github.com/Shopify/kubeaudit/issues/343
The following changes were done:

- `seccomp` auditor changed to scan `securityContext` in pod and containers to find `seccompProfile`. :warning:
- New `fix` created in `seccomp` auditor to create seccomp profile in a pod and containers and to remove profile in containers. The auditor updated to use the new `fix`.
- `SeccompAnnotationMissing` rule renamed to `SeccompProfileMissing`. :warning:
- `SeccompDeprecatedContainer` rule dropped. :warning:
- Tests updated to cover changes in auditor and fix.
- Documentation updated.

Fixes #343

##### Type of change

- [X] This change requires a documentation update :book:
- [X] Breaking changes :warning:

##### How Has This Been Tested?

- Mostly by automated tests.
- I've also executed kubeaudit for examples from documentation to update them.

##### Checklist:

- [ ] I have :tophat: my changes (A 🎩 specifically includes pulling down changes, setting them up, and manually testing the changed features and potential side effects to make sure nothing is broken)
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] The test coverage did not decrease
- [X] I have signed the appropriate [Contributor License Agreement](https://cla.shopify.com/)

### CODE REVIEW

**Comment by dani-santos-code** [comment]
hi, @Ser87ch ! Thank you for your contribution! I will take a closer look at your PR either today or tomorrow. In the meantime, would you mind adding support to both annotations and `seccompProfile`?

**Comment by Ser87ch** [comment]
@dani-santos-code Thanks for your reply. May I ask, why? Seccomp annotation support will be removed quite soon. 
Additionally, supporting both will add more complications, like what to do if we've got both security context and annotations set.

**Comment by Ser87ch** [comment]
Thanks @dani-santos-code , I'll let you know tomorrow.

**Comment by genevieveluyt** [comment]
If the annotation is going to stop working and the security profile has already been available for some time, I think dropping annotation support is fair. +1 to @dani-santos-code's suggestion though, if it's not too much of a hassle.

I think ideally `Fix` would remove the seccomp annotation as well, if it's present.

**Comment by Ser87ch** [comment]
I've encountered an issue, that I don't know how to fix yet. I'm testing a change that you suggested with the following pod spec:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod
  namespace: seccomp-profile-missing-annotations
  annotations:
    seccomp.security.alpha.kubernetes.io/pod: runtime/default
    container.seccomp.security.alpha.kubernetes.io/container: localhost/bla
spec:
  containers:
    - name: container
      image: scratch
```
The problem is when I fetch security context the pod spec, it returns me a security context with seccomp profile set:
```
&PodSecurityContext{SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,SupplementalGroups:[],FSGroup:nil,RunAsGroup:nil,Sysctls:[]Sysctl{},WindowsOptions:nil,FSGroupChangePolicy:nil,SeccompProfile:&SeccompProfile{Type:RuntimeDefault,LocalhostProfile:nil,},}
```
I have a feeling that k8s library returns seccomp profile in security context even if only seccomp annotation set. If it's the case I'm not able to differentiate between seccomp set in security context and annotations.
I've committed this change, this test fails `seccomp-profile-missing-disabled-container.yaml`. Am I missing something?

**Comment by Ser87ch** [comment]
It's not a k8s library problem, it the way k8s API works. If I export the pod spec I mentioned in the previous comment by using `kubectl get pod pod -n seccomp-profile-missing-annotations -o yaml`, it returns:
```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    container.seccomp.security.alpha.kubernetes.io/container: localhost/bla
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{"container.seccomp.security.alpha.kubernetes.io/container":"localhost/bla","seccomp.security.alpha.kubernetes.io/pod":"runtime/default"},"name":"pod","namespace":"seccomp-profile-missing-annotations"},"spec":{"containers":[{"image":"scratch","name":"container"}]}}
    seccomp.security.alpha.kubernetes.io/pod: runtime/default
  creationTimestamp: "2022-09-13T08:46:02Z"
  name: pod
  namespace: seccomp-profile-missing-annotations
  resourceVersion: "1735"
  uid: cc1671c0-bbde-4a98-91ee-5d4761cb19d4
spec:
  containers:
  - image: scratch
    imagePullPolicy: Always
    name: container
    resources: {}
    securityContext:
      seccompProfile:
        localhostProfile: bla
        type: Localhost
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-jh5xb
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: kubeaudit-test-control-plane
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
 ...
```
So, I don't believe it's possible to catch a situation when seccompProfile is absent in securityContext but is present in annotations. 
@genevieveluyt @dani-santos-code what do you think? Should I just raise a warning if seccomp annotations are present?

**Comment by genevieveluyt** [comment]
Based on [these docs](https://kubernetes.io/docs/tutorials/security/seccomp/#enable-the-use-of-runtimedefault-as-the-default-seccomp-profile-for-all-workloads), it sounds like having seccomp enabled is the new default behaviour, so kubeaudit might not even have to check that seccomp is enabled but rather should produce an error if seccomp is explicitly disabled (eg. if the profile is set to Unconfined). Did you run your kubelet with the `--seccomp-default` command line flag? I wonder what the API returns for the seccompProfile if you don't use that flag?

**Comment by Ser87ch** [comment]
@genevieveluyt thanks for your comment. As far as I understand from the [link you shared](https://kubernetes.io/docs/tutorials/security/seccomp/#enable-the-use-of-runtimedefault-as-the-default-seccomp-profile-for-all-workloads) and [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/), it is a default behaviour but only since version 1.25 (which is in beta). I tested it with the default `v1.20.15`, where this behaviour isn't even available.

**Comment by Ser87ch** [comment]
> I see. If neither the annotation nor the seccomp profile are set, what does the library return for the profile?

There is misunderstanding here. Annotation is set in [the example](https://github.com/Shopify/kubeaudit/pull/475#issuecomment-1244080138) I put above. The problem is that Kubernetes API returns also seccomp profile in the security context when only annotation is set.

**Comment by Ser87ch** [comment]
If neither is set, then it returns empty securityContext.

**Review by genevieveluyt — COMMENTED** [review]
I think we should check for annotations separately (an auditor can return multiple results).

1. Check the pod for seccomp annotations. If there are any, produce a `warning` result where the pending fix removes all said annotations
2. Check the pod and containers for seccomp profile (as you do currently). If it's unset or set to Unconfined, produce an `error` result and the pending fix sets the profile to RuntimeDefault

This means in manifest mode you might get both a warning and an error, but I think that's ok. In cluster and local mode, since autofix doesn't do anything, likely someone would manually fix the annotation warning by removing annotation, rerun kubeaudit, and then get the seccomp profile error. In future we could upgrade the annotation warning to an error results as well.

What do you think?

**Comment by Ser87ch** [comment]
Thanks, your proposal sounds good. I will change accordingly.

**Comment by genevieveluyt** [comment]
Well that's a rather unexpected and annoying way for the API to behave lol. There's no way in kubeaudit to only run a check manifest mode, no. What if, for number 1 above ie

> 1. Check the pod for seccomp annotations. If there are any, produce a warning result where the pending fix removes all said annotations

We change that check to "the pod has _only_ seccomp annotations"? That should only ever happen in manifest mode then right? I don't think there's anything we can do about the annotations in cluster and local mode given how the API behaves...

**Comment by Ser87ch** [comment]
@genevieveluyt I've introduced a separate method that audit annotations only if seccomp profile is missing for both pod and all containers. This way it will be easier to remove this logic once seccomp annotation support is dropped.

File: auditors/seccomp/seccomp.go [hunk 1]
```diff
@@ -66,164 +58,143 @@ func (a *Seccomp) Audit(resource k8s.Resource, _ []k8s.Resource) ([]*kubeaudit.A
...
+	if podSpec == nil {
+		return nil
+	}
 
-	if isSeccompAnnotationMissing(annotations, PodAnnotationKey) {
-		// If all the containers have container-level seccomp annotations then we don't need a pod-level annotation
-		if isSeccompEnabledForContainers(annotations, resource) {
+	if isPodSeccompProfileMissing(podSpec.SecurityContext) {
+		// If all the containers have container-level seccomp profiles then we don't need a pod-level profile
+		if isSeccompEnabledForAllContainers(resource) {
```
[the remark this answers is not in the document]
Comment (Ser87ch) [reply auditors/seccomp/seccomp.go:107]: Annotations are checked in a separate method now.

**Comment by Ser87ch** [comment]
@genevieveluyt could you review PR?

**Comment by genevieveluyt** [comment]
@Ser87ch thank you for making the changes, I will review and 🎩 by end of next week.

File: auditors/seccomp/fixtures/seccomp-disabled-localhost.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+  namespace: seccomp-disabled-localhost
+spec:
+  securityContext:
+        seccompProfile:
```
[the remark this answers is not in the document]
Comment (Ser87ch) [reply auditors/seccomp/fixtures/seccomp-disabled-localhost.yml]: fixed

**Review by genevieveluyt — APPROVED** [review]
🎩 looks good! Sorry for the delays, thanks so much for your contribution!

**Review by dani-santos-code — APPROVED** [review]
I also tested on my end and it works really well! Thanks for your changes @Ser87ch ! 🙏

**Comment by Ser87ch** [comment]
@dani-santos-code thanks, how can we merge this one first?

**Pull request merged**